Bug: [no-unnecessary-type-assertion] false positive on `reduce` initial value behind optional chaining

### Before You File a Bug Report Please Confirm You Have Done The Following...

- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues) and found none that matched my issue.
- [x] I have [tested with the latest version](https://typescript-eslint.io/play) and reproduced the issue.

### Playground Link

https://typescript-eslint.io/play/#ts=5.9.3&sourceType=module&code=CYUwxgNghgTiAEYD2A7AzgF3gSwyAtmgFzyYzYoDmA2gLrwA+8AriqAGYUjADcAUHwD0g+ADFolSt3hQ0LFCnAg0aWAE8ANPABGzLBgAWCMLKzY5AJQCiARQCqASWsARErlIhucwwihgwzPjM0BhIMEIiUorkYPAADrBQ+CB4MPBI7PAABnDAzGAgWQB08ADquAZIejgYWll+YFk4KOwgMHKy2QDeAL7FfMjoWMisGHIAvDUEaAD8Rbn5IAAUSw1auAQAlPDjAHzwXXzwMv7UG-j0k6un5-QzM-AADNsA1PAAjPzHcBjMMCgnMD8HpaXoySzgMLAAA8ZAolC0KEC2jau02-Ai8AAwqgMDBTCQfPAAMoAQQAslZ4AsCvAAO4VKpYJBxDDYVBQCCIAxQCjwnByAByAHkACrwdgSKTAIp8UCQWDGVCYKaEABMJDhVDo-EGKpGKDGap2qrQavm3EWKzWqu2ewOR0BZzwFxN1zAzoIdwez3gb0+jp+fwBDWBoJ64PgFkhMBhWoR8CR+BRMDR-CAA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tiacTJTIAhtEK0ipWkOTJE0fJQ5N0UOdA7RI4MAF8QOoA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eFYDAruuGAL4g9A

### Repro Code

```ts
declare const items: string[] | undefined;

// Flagged as unnecessary, but the cast is REQUIRED: it seeds the accumulator
// generic parameter of `reduce`. Without it, `acc` infers as `{}`.
const counts = items?.reduce((acc, item) => {
  acc[item] = (acc[item] ?? 0) + 1;
  return acc;
}, {} as Record<string, number>);

// Contrast: the SAME reduce WITHOUT optional chaining is NOT flagged.
declare const items2: string[];
const counts2 = items2.reduce((acc, item) => {
  acc[item] = (acc[item] ?? 0) + 1;
  return acc;
}, {} as Record<string, number>);
```

### ESLint Config

```json
{
  "rules": {
    "@typescript-eslint/no-unnecessary-type-assertion": "error"
  }
}
```

### tsconfig

```json
{
  "compilerOptions": {
    "strict": true
  }
}
```

### Expected Result

No report on line 8. The `{} as Record<string, number>` assertion is **not** unnecessary — it seeds the generic accumulator type parameter (`U`) of `Array.prototype.reduce<U>(cb, initialValue: U)`. Removing it makes `U` infer as `{}`, so the callback body no longer compiles:

```
error TS7053: Element implicitly has an 'any' type because expression of type
'string' can't be used to index type '{}'.
```

Because removing the assertion changes the program's types (and breaks compilation), it is load-bearing and should not be flagged.

### Actual Result

The rule reports on the `reduce` initial value:

> This assertion is unnecessary since the receiver accepts the original type of the expression

Note it is flagged **only when `reduce` is called behind an optional chain** (`items?.reduce(...)`). The identical `reduce` on a non-nullable array (`items2.reduce(...)`, the contrast case) is correctly **not** flagged. So the false positive is specific to how the contextual/receiver type of the `initialValue` argument is resolved through an optional-chained call: `{}` is assignable to `Record<string, number>`, so the "receiver accepts the original type" check passes — but that check uses the type parameter as already inferred *with* the assertion, and doesn't account for the assertion being what pins that inference.

### Additional Info

- This is the same "unnecessary from the perspective of assignment, but required for inference" class as #12250 (fixed for phantom type arguments in v8.59.1 via #12269). That fix does **not** cover this variant: a `reduce` initial-value assertion surfaced through **optional chaining** still reproduces on the latest release.
- Reproduces on **@typescript-eslint/eslint-plugin 8.62.1** with **TypeScript 5.9.3 and 6.0.3**, ESLint 10.6.0.

--- a/packages/eslint-plugin/src/rules/no-unnecessary-type-assertion.ts
+++ b/packages/eslint-plugin/src/rules/no-unnecessary-type-assertion.ts
@@ -555,11 +555,9 @@ export default createRule<Options, MessageIds>({
       }
       // An optional-chained callee (`foo?.bar(...)`) types as `<method> | undefined`,
       // and a union exposes no call signatures — strip the nullability first.
-      const calleeType = checker.getNonNullableType(
-        checker.getTypeAtLocation(
-          services.esTreeNodeToTSNodeMap.get(parent.callee),
-        ),
-      );
+      const calleeType = services
+        .getTypeAtLocation(parent.callee)
+        .getNonNullableType();
       const signatures = calleeType.getCallSignatures();
       if (signatures.length <= 1) {
         return false;

--- a/packages/eslint-plugin/src/rules/no-unnecessary-type-assertion.ts
+++ b/packages/eslint-plugin/src/rules/no-unnecessary-type-assertion.ts
@@ -553,8 +553,12 @@ export default createRule<Options, MessageIds>({
       ) {
         return false;
       }
-      const calleeType = checker.getTypeAtLocation(
-        services.esTreeNodeToTSNodeMap.get(parent.callee),
+      // An optional-chained callee (`foo?.bar(...)`) types as `<method> | undefined`,
+      // and a union exposes no call signatures — strip the nullability first.
+      const calleeType = checker.getNonNullableType(
+        checker.getTypeAtLocation(
+          services.esTreeNodeToTSNodeMap.get(parent.callee),
+        ),
       );
       const signatures = calleeType.getCallSignatures();
       if (signatures.length <= 1) {

--- a/packages/eslint-plugin/tests/rules/no-unnecessary-type-assertion.test.ts
+++ b/packages/eslint-plugin/tests/rules/no-unnecessary-type-assertion.test.ts
@@ -2876,9 +2876,6 @@ fn1(() => {
       ],
       output: 'const foo = () => ({ lol: 123 as number }.lol) + 54321;',
     },
-    // optional-chained call to a non-overloaded callee: assertions that are
-    // genuinely unnecessary must still be reported
-    // https://github.com/typescript-eslint/typescript-eslint/issues/12485
     {
       code: `
 declare const maybeFn: ((arg: string | number) => void) | undefined;

--- a/packages/eslint-plugin/tests/rules/no-unnecessary-type-assertion.test.ts
+++ b/packages/eslint-plugin/tests/rules/no-unnecessary-type-assertion.test.ts
@@ -900,6 +900,30 @@ const test = inferred({
 
 console.log(test.options.parameters.potato);
     `,
+    // https://github.com/typescript-eslint/typescript-eslint/issues/12485
+    `
+declare const items: string[] | undefined;
+
+const counts = items?.reduce(
+  (acc, item) => {
+    acc[item] = (acc[item] ?? 0) + 1;
+    return acc;
+  },
+  {} as Record<string, number>,
+);
+    `,
+    `
+declare const o:
+  | {
+      fn<U>(g: (memo: U) => U, initial: U): U;
+      fn<T>(g: (memo: T) => T): T | undefined;
+    }
+  | undefined;
+enum E {
+  A = 1,
+}
+const x: E | undefined = o?.fn(n => n | 0, 0 as E);
+    `,
   ],
 
   invalid: [
@@ -2851,6 +2875,28 @@ fn1(() => {
         },
       ],
       output: 'const foo = () => ({ lol: 123 as number }.lol) + 54321;',
+    },
+    // optional-chained call to a non-overloaded callee: assertions that are
+    // genuinely unnecessary must still be reported
+    // https://github.com/typescript-eslint/typescript-eslint/issues/12485
+    {
+      code: `
+declare const maybeFn: ((arg: string | number) => void) | undefined;
+declare const s: string;
+maybeFn?.(s as string | number);
+      `,
+      errors: [
+        {
+          column: 11,
+          line: 4,
+          messageId: 'contextuallyUnnecessary',
+        },
+      ],
+      output: `
+declare const maybeFn: ((arg: string | number) => void) | undefined;
+declare const s: string;
+maybeFn?.(s);
+      `,
     },
   ],
 });

